paginate by startkey

aogier opened the pull request:
<!--
Thanks for your hard work, please ensure all items are complete before opening.
-->
## Checklist

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](../blob/master/DCO1.1.txt)
- [x] Added tests for code changes _or_ test/build only changes
- [x] Updated the change log file (`CHANGES.md`|`CHANGELOG.md`) _or_ test/build only changes
- [x] Completed the PR template below:

## Description

Make `Result`'s iteration faster paginating by `startkey` instead of `skip`.

Fixes #436 

## Approach

As per #436 , paginating by `skip` could be super-slow. This patch fix the problem by using `startkey` so every iteration run in constant time.

## Schema & API Changes

- "No change"

## Security and Privacy

- "No change"

## Testing

- No new tests because current ones are enough

## Monitoring and Logging

- "No change"

smithsz commented on src/cloudant/result.py:
```diff
@@ -347,24 +347,27 @@ def __iter__(self):
+                last = result.pop()
                 for row in result:
                     yield row
                 if len(result) < self._page_size:
```
Change to `if len(result) < self._page_size + 1`?

Also, I think a comment here would be good. Something like:

> We expect `len(result) = self._page_size + 1` results, if we have ` self._page_size` or less it means we are on the last page and need to return the last result.

aogier replied:
you're right, this way we save a hit on the worst case when ndocs%pagesize==0 ... neat !

aogier commented:
I've found a bug while implementing comments requests: when a view's key is an integer, view does not return because startkey='' apparently want to cast to string:

```sh
# this does work
c /temp-cloudant/_design/dd/_view/vv\?limit=3                
{
  "total_rows": 6,
  "offset": 0,
  "rows": [
    {
      "id": "f41a9664608b060ce1a488ed96005a35",
...

# this does not

c /temp-cloudant/_design/dd/_view/vv\?limit=3\&startkey\='""'
{
  "total_rows": 6,
  "offset": 6,
  "rows": []
}

```

aogier commented:
Ok I've changed a bit the approach so we use `startkey` in GET if and only if user specifies one. I've implemented this via a partial object (I like them so much, so elegant). Also, added the comment and avoid an empty hit if we are at the view's end!

aogier commented:
I'm working on tests to cover the integer key corner case

aogier commented:
added a test that would not have passed in 19a4abe078f9ba55083ed44fc5baddf18b45437c but pass now in 15d216d46fb68a0f6e7b43089a73a5a6d4324996 ... reverse TDD :+1:

aogier commented:
have to write more tests

aogier commented:
ok I've added another test for observing pagination behaviour and now everything seems pretty stable and robust to me. @smithsz thank you for letting me know what you think :)

ciao

smithsz commented on src/cloudant/result.py:
```diff
@@ -327,7 +328,7 @@ def _handle_result_by_key_slice(self, key_slice):
         """
         Provides iteration support, primarily for large data collections.
-        The iterator uses the ``skip`` and ``limit`` options to consume
+        The iterator uses the ``startkey`` and ``limit`` options to consume
```
and ``startkey_docid``

aogier replied:
ah ok, I've left it back on purpose viewing it as an "hidden feature" but I'll amend docs

smithsz commented on tests/unit/result_tests.py:
```diff
@@ -643,5 +645,43 @@ def test_iteration_no_data(self):
+        """
+        Test that iteration works as expected when keys are integer.
+        """
+        result = Result(self.view007)
```
can you put a small page size in here? maybe `page_size=10`

aogier replied:
sure

smithsz commented on tests/unit/result_tests.py:
```diff
@@ -643,5 +645,43 @@ def test_iteration_no_data(self):
+                self.outer = outer
+
+            def call(self, *args, **kwargs):
+                self.outer.assertNotEqual((args, kwargs), self.last_call,
```
This is just asserting that the current view query isn't the same as the previous right? How about checking against a list of expected stuff, i.e.
```
call 1: {'limit': 28}
call 2: {'limit': 28, 'startkey': 1, 'startkey_docid': 'julia027'}
call 3: {'limit': 28, 'startkey': 1, 'startkey_docid': 'julia054'}
call 4: {'limit': 28, 'startkey': 1, 'startkey_docid': 'julia081'}
```

aogier replied:
It's indeed more strict, I like it

smithsz commented:
This is a great. I've been doing some testing with other options (i.e. descending=True) and everything works nicely.
I notice there's a check in there for `limit` and `skip` params not being supported. I guess this is still true for `limit` but we should be able to use `skip` now right? It'll want removing from the query params after the first call though. What do you think?

aogier commented:
Seems a good idea to me, it opens up to more use cases and it's totally manageable in the current implementation. 

BTW, I think we could also reason about allowing `limit`: we use `limit` in the request context for efficiently paginating over a result set, but from the user standpoint a `limit` clause mean "I want to iterate over this set and get at most N total results" and this could totally be feasible by servin count(results) < limit. What do you think? Just a couple cents about another iteration eh :)

I'll begin to implement for a skip param!

smithsz commented on src/cloudant/result.py:
```diff
@@ -347,21 +348,47 @@ def __iter__(self):
             raise ResultException(104, self._page_size)
 
-        skip = 0
+        initial_options = {option: value
```
Not sure about this. Bit tricky to read. How about...
```py
        _call = partial(self._ref, limit=self._page_size + 1, **self.options)

        init_opts = {
            'skip': self.options.pop('skip', None),
            'startkey': self.options.pop('startkey', None)
        }

        response = _call(**{k: v for k, v in init_opts.items() if v is not None})
```

aogier replied:
ok, it's more concise and definitely easier to read!

aogier commented:
did the last changes and also added a changelog entry!

aogier commented:
queries are to be paginated via `bookmark`, I'm adding more tests

aogier commented:
ok so basically I was abusing nose's filtering options and skipped query tests, now there is a possibile implementation on which I decided to stay straight and implement a specific inner method for queries and views as they are so different in schema. Thank you, regards

emlaver commented on tests/unit/query_result_tests.py:
```diff
@@ -428,21 +428,11 @@ def test_iteration_with_invalid_options(self):
         Test that iteration raises an exception when "skip" and/or "limit" are
```
Need to fix this comment now that tests with skip are removed.

aogier replied:
you're right!

emlaver commented:
For tests to run under Python 3, this will need to be rebased to include #438 which fixes the `pylint` warning.

aogier commented:
ok I've fixed test docstring and rebased on 2eb1261838d02b5ee63d4a572095a78bcfcee825, were jenkins failing because of pylint?

emlaver commented:
@aogier Yes, the Python 3 tests were showing the pylint warning.

emlaver commented on CHANGES.md:
```diff
@@ -1,5 +1,6 @@
 # Unreleased
 
+- [IMPROVED] Faster `Result` iterator.
```
We should probably give some detail and not use 'Faster'
i.e. `Improve Result iteration by paginating with 'startkey'`.  @smithsz wdyt?

aogier commented:
Great !

I've squashed it and modified the changelog entry:
```
- [IMPROVED] Updated `Result` iteration by paginating with views' `startkey` and
  query's `bookmark`.
```

Thank you, regards
